builtin/providers/aws/resource_aws_codebuild_project.go

```diff
@@ -0,0 +1,722 @@
+
+			for _, envVariablesConfig := range envVariables {
+				config := envVariablesConfig.(map[string]interface{})
+
+				projectEnvironmentVar := &codebuild.EnvironmentVariable{}
+
+				if v := config["name"].(string); v != "" {
+					projectEnvironmentVar.Name = &v
+				}
+
+				if v := config["value"].(string); v != "" {
+					projectEnvironmentVar.Value = &v
+				}
+
+				projectEnvironmentVariables = append(projectEnvironmentVariables, projectEnvironmentVar)
+			}
+
+			projectEnv.EnvironmentVariables = projectEnvironmentVariables
+		}
+	}
+
+	return projectEnv
+}
+
+func expandProjectSource(d *schema.ResourceData) codebuild.ProjectSource {
+	configs := d.Get("source").(*schema.Set).List()
+	projectSource := codebuild.ProjectSource{}
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		sourceType := data["type"].(string)
+		location := data["location"].(string)
+		buildspec := data["buildspec"].(string)
+
+		projectSource = codebuild.ProjectSource{
+			Type:      &sourceType,
+			Location:  &location,
+			Buildspec: &buildspec,
+		}
+
+		if v, ok := data["auth"]; ok {
+			if len(v.(*schema.Set).List()) > 0 {
+				auth := v.(*schema.Set).List()[0].(map[string]interface{})
+
+				projectSource.Auth = &codebuild.SourceAuth{
+					Type:     aws.String(auth["type"].(string)),
+					Resource: aws.String(auth["resource"].(string)),
+				}
+			}
+		}
+	}
+
+	return projectSource
+}
+
+func resourceAwsCodeBuildProjectRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codebuildconn
+
+	log.Printf("[DEBUG] CodeBuild Read Project: %s", d.Id())
+
+	resp, err := conn.BatchGetProjects(&codebuild.BatchGetProjectsInput{
+		Names: []*string{
+			aws.String(d.Id()),
+		},
+	})
+
+	if err != nil {
+		return fmt.Errorf("[ERROR] Error retreiving Projects: %q", err)
+	}
+
+	// if nothing was found, then return no state
+	if len(resp.Projects) == 0 {
+		log.Printf("[INFO]: No projects were found, removing from state")
+		d.SetId("")
+		return nil
+	}
+
+	project := resp.Projects[0]
+
+	artifacts := flattenAwsCodebuildProjectArtifacts(project.Artifacts)
+	if artifacts != nil {
+		d.Set("artifacts", artifacts)
+	}
+
+	if err := d.Set("environment", schema.NewSet(resourceAwsCodeBuildProjectEnvironmentHash, flattenAwsCodebuildProjectEnvironment(project.Environment))); err != nil {
+		return err
+	}
+
+	if err := d.Set("source", flattenAwsCodebuildProjectSource(project.Source)); err != nil {
+		return err
+	}
+
+	d.Set("description", project.Description)
+	d.Set("encryption_key", project.EncryptionKey)
+	d.Set("name", project.Name)
+	d.Set("service_role", project.ServiceRole)
+	d.Set("timeout", project.TimeoutInMinutes)
+
+	if err := d.Set("tags", tagsToMapCodeBuild(project.Tags)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceAwsCodeBuildProjectUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codebuildconn
```

Comment:
Do we want to update only the pieces that need to change or pass the entire UpdateProjectInput every time? In an Update func, we usually only send the pieces that change by using the if d.HasChange("")

Can this pattern be used here too?